skfp/descriptors/topological.py

```diff
@@ -0,0 +1,210 @@
+
+    Parameters
+    ----------
+    mol : RDKit Mol object
+        The molecular graph for which the Graph Distance Index is to be calculated.
+
+    distance_matrix : np.ndarray, optional
+        Precomputed distance matrix. If not provided, it will be calculated using RDKit.
+
+    References
+    ----------
+    .. [1] PyChem documentation
+        <https://storage.googleapis.com/google-code-archive-downloads/v2/code.google.com/pychem/manual.pdf>
+
+    Examples
+    --------
+    >>> from rdkit import Chem
+    >>> from descriptors.topological import graph_distance_index
+    >>> mol = Chem.MolFromSmiles("C1=CC=CC=C1")  # Benzene
+    >>> graph_distance_index(mol)
+    261.0
+"""
+    if distance_matrix is None:
+        distance_matrix = GetDistanceMatrix(mol)
+    distances = distance_matrix[np.triu_indices_from(distance_matrix, k=1)]
+    distance_counts = Counter(distances)
+    return sum((k * f) ** 2 for k, f in distance_counts.items())
+
+
+def zagreb_index(mol: Mol, adjacency_matrix: Optional[np.ndarray] = None) -> int:
+    """
+    Zagreb Index.
+
+    The implementation uses RDKit to calculate the first Zagreb index for molecular graphs.
+    It is defined as the sum of the squares of the degrees of all atoms in the molecule.
+
+    Parameters
+    ----------
+    mol : RDKit Mol object
+        The molecular graph for which the Zagreb index is to be calculated.
+
+    adjacency_matrix : np.ndarray, optional
+        Precomputed adjacency matrix. If not provided, it will be calculated using RDKit.
+
+    References
+    ----------
+    .. [1] Gutman, Ivan.
+        "Degree-based topological indices."
+        Croatica chemica acta 86.4 (2013): 352.
+        <https://hrcak.srce.hr/file/166451>
+
+    Examples
+    --------
+    >>> from rdkit import Chem
+    >>> from descriptors.topological import zagreb_index
+    >>> mol = Chem.MolFromSmiles("C1=CC=CC=C1")  # Benzene
+    >>> zagreb_index(mol)
+    24
+    """
+    if adjacency_matrix is None:
+        adjacency_matrix = rdmolops.GetAdjacencyMatrix(mol)
+
+    degree_sequence = adjacency_matrix.sum(axis=1)
+    return sum(degree_sequence ** 2)
+
+
+def polarity_number(mol: Mol, distance_matrix: Optional[np.ndarray] = None, carbon_only: bool = False) -> int:
+    """
+    Polarity Number (Pol).
+
+    The implementation calculates the Polarity Number of a molecule.
+    It is defined as the total number of unordered pairs of vertices (atoms)
+    in a molecular graph that are separated by a graph distance of exactly 3.
+
+    The polarity number provides information about the structural connectivity of a molecule.
+
+    Parameters
+    ----------
+    mol : RDKit Mol object
+        The molecular graph for which the Polarity Number is to be calculated.
+
+    distance_matrix : np.ndarray, optional
+        Precomputed distance matrix. If not provided, it will be calculated using RDKit.
+
+    carbon_only : bool, default=False
```

Comment:
Edge case - what if molecules does not have any carbons? This should raise an explicit error then. You can check this, i.e. if `atom_indices` is empty or not. Also, write a test for this.